examples/examples/core/containers.py

```diff
@@ -0,0 +1,202 @@
+"""
+Using Container classes
+=======================
+
+``ctapipe.core.Container`` is the base class for all event-wise data
+classes in ctapipe. It works like a object-relational mapper, in that it
+defines a set of ``Fields`` along with their metadata (description,
+unit, default), which can be later translated automatially into an
+output table using a ``ctapipe.io.TableWriter``.
+
+"""
+
+from functools import partial
+
+import numpy as np
+from astropy import units as u
+
+from ctapipe.core import Container, Field, Map
+
+######################################################################
+# Let’s define a few example containers with some dummy fields in them:
+# 
+
+class SubContainer(Container):
+    junk = Field(-1, "Some junk")
+    value = Field(0.0, "some value", unit=u.deg)
+
+
+class TelContainer(Container):
+    # defaults should match the other requirements, e.g. the defaults
+    # should have the correct unit. It most often also makes sense to use
+    # an invalid value marker like nan for floats or -1 for positive integers
+    # as default
+    tel_id = Field(-1, "telescope ID number")
+
+    # For mutable structures like lists, arrays or containers, use a `default_factory` function or class
+    # not an instance to assure each container gets a fresh instance and there is no hidden
+    # shared state between containers.
+    image = Field(default_factory=lambda: np.zeros(10), description="camera pixel data")
+
+
+class EventContainer(Container):
+    event_id = Field(-1, "event id number")
+
+    tels_with_data = Field(
+        default_factory=list, description="list of telescopes with data"
+    )
+    sub = Field(
+        default_factory=SubContainer, description="stuff"
+    )  # a sub-container in the hierarchy
+
+    # A Map is like a defaultdictionary with a specific container type as default.
+    # This can be used to e.g. store a container per telescope
+    # we use partial here to automatically get a function that creates a map with the correct container type
+    # as default
+    tel = Field(default_factory=partial(Map, TelContainer), description="telescopes")
+
+
+######################################################################
+# Basic features
+# --------------
+# 
+
+ev = EventContainer()
+
+
+######################################################################
+# Check that default values are automatically filled in
+# 
+
+print(ev.event_id)
+print(ev.sub)
+print(ev.tel)
+print(ev.tel.keys())
+
+# default dict access will create container:
+print(ev.tel[1])
+
+
+######################################################################
+# print the dict representation
+# 
+
+print(ev)
+
+
+######################################################################
+# We also get docstrings “for free”
+# 
+
+?EventContainer
```

Comment:
This is just invalid python syntax, which is why black complains.

Reply:
Not sure how to solve this... thinking about it, it's clear that we loose most jupyter / ipython features in this conversion, like the `<class>?` call to open the docstrins.

Maybe replace with `help(EventContainer)`? That works in both python and jupyter

Reply:
We will also loose commands like `!ls`

Reply:
Yes I think so too. I'm gonna go ahead and replace everything that fails with a python equivalent

Reply:
yes, it's ok to replace these. Also `EventSource?` is just `help(EventSource)` in standard python

Reply:
https://jupytext.readthedocs.io/en/latest/advanced-options.html#magic-commands

Reply:
It's only a problem for when a user exports to a py script,  which I think is ok. 

The magic commands should still work in Jupyter, and I guess also in the documentation?  Or does it not use Jupyter to render the pages?

Reply:
But we now have the notebooks as `.py` scripts in git and we try to run black on them, which chokes on the invalid syntax...

Reply:
which results in the docs showing the command as a comment too

Reply:
Take a look here maybe, at the end of the page
https://ctapipe--2377.org.readthedocs.build/en/2377/tutorials/ctapipe_overview.html